[GSoC-Phase 1]JSON Parser in Python and Tests

jiachengxu opened the pull request:
This PR contains the work of phase 1 of GSoC.
- Add a JSON parser in Python to parse the json dataset.
- Add tests that can be reused one the CPP JSON parser is implemented.

yongtang commented on tensorflow_io/json/python/ops/json_ops.py:
```diff
@@ -0,0 +1,89 @@
+# JSON parser in Python for testing.
+
+
+def JSONParser(filenames):
```
@jiachengxu The name should be changed to `jsonParser`.

yongtang commented on tensorflow_io/json/python/ops/json_ops.py:
```diff
@@ -0,0 +1,89 @@
+def JSONFeatureDataset(filenames):
+  jsondataset = []
+  jsondataset = jsonParser(filenames)
+  return tf.data.Dataset.from_tensor_slices(jsondataset)
```
@jiachengxu  `JSONFeatureDataset` and `JSONLabelDataset` should be just one. I don't think we need two Dataset for this case.

yongtang commented:
@jiachengxu we only need on JSONDataset. You can use this one Dataset to open feature and label files. The zip op is directly available from tf.data.Dataset.zip((dataset1, dataset2)) and there is no need to add a wrapper for it.

jiachengxu commented:
Hi @yongtang ,
Thanks for your comments.

> You can use this one Dataset to open feature and label files.
I am also thinking about what JSON schema should the dataset follows, I think there are 2 options:
- features and labels are in the same JSON file, i.e., the file should look like this:
```
[{
	"sample1": {
		"features": [{
				"feature1": "featurevalue1"
			},
			{
				"feature2": "featurevalue2"
			}
		],
		"labels": [{
				"label1": "labelvalue1"
			},
			{
				"label2": "labelvalue2"
			}
		]
	},
	"sample2": {
		"features": [{
				"feature1": "featurevalue1"
			},
			{
				"feature2": "featurevalue2"
			}
		],
		"labels": [{
				"label1": "labelvalue1"
			},
			{
				"label2": "labelvalue2"
			}
		]
	}
}]
```
Another option would be put features and labels into 2 files, like:
```
[{
	"sample1": {
		"features": [{
				"feature1": "featurevalue1"
			},
			{
				"feature2": "featurevalue2"
			}
		]
	},
	"sample2": {
		"features": [{
				"feature1": "featurevalue1"
			},
			{
				"feature2": "featurevalue2"
			}
		]
	}
}]
```
```
[{
	"sample1": {
		"labels": [{
				"label1": "labelvalue1"
			},
			{
				"label2": "labelvalue2"
			}
		]
	},
	"sample2": {
		"labels": [{
				"label1": "labelvalue1"
			},
			{
				"label2": "labelvalue2"
			}
		]
	}
}]
```
Which you would you perfer?

> The zip op is directly available from tf.data.Dataset.zip((dataset1, dataset2)) and there is no need to add a wrapper for it.

Yes, that's true, I did that because if I implement the JSON parser in CPP, and if I have both LabelDataset and FeatureDataset, I need to have the wrapper, right? I saw the [mnist support](https://github.com/tensorflow/io/blob/master/tensorflow_io/mnist/python/ops/mnist_ops.py#L62-L65) did that.

Thanks,
Jiacheng

jiachengxu commented:
Hi @yongtang, now I see your point, I think I can just do the following, just put the features and labels to the same array:
```
[{
	"sample1": [{
			"feature1": "featurevalue1"
		},
		{
			"feature2": "featurevalue2"
		},
		{
			"label1": "labelvalue1"
		},
		{
			"label2": "labelvalue2"
		}
	],
	"sample2": [{
			"feature1": "featurevalue1"
		},
		{
			"feature2": "featurevalue2"
		},
		{
			"label1": "labelvalue1"
		},
		{
			"label2": "labelvalue2"
		}
	]
}]
```
This would be more flexible, for example, if user has a file that contains both features and labels, the user can pass it to JSONDataset, and specify input shapes when passing it to keras, and if the user has features and labels in different files, the user can still pass them into JSONDataset to get 2 datasets, and then zip them. WDYT?

yongtang commented:
@jiachengxu Yes you can basically consider JSON as columnized dataset. It is falls into the same category as Parquet, Arrow, Avro, CSV, etc. You could extract columns so that you have label and feature from the same dataset and pass it to tf.keras, or combine two dataset into one with zip to pass it to tf.keras. See slightly off-topic discussion in #315.

yongtang commented:
@jiachengxu any update? It would be great to have a simple test setup (with python only json implementation) that runs and passes Travis CI.

Once we have a basic setup, the next step of building with jsoncpp will be much easier.

jiachengxu commented:
Hi @yongtang , I have updated the PR, and the Linux builds and releases can pass, but the OS X ones are always falling, do you have any ideas about that?

yongtang commented:
@jiachengxu The error is available from Travis CI log:
```
==================================== ERRORS ====================================
_____________________ ERROR collecting tests/test_json.py ______________________
ImportError while importing test module '/Users/travis/build/tensorflow/io/tests/test_json.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
test_json.py:23: in <module>
    import tensorflow_io.json as json_io
/Library/Python/2.7/site-packages/tensorflow_io/json/__init__.py:27: in <module>
    from tensorflow_io.json.python.ops.json_ops import JSONFeatureDataset
E   ImportError: cannot import name JSONFeatureDataset
=============================== warnings summary ===============================
/Library/Python/2.7/site-packages/pandas/util/testing.py:22
  /Library/Python/2.7/site-packages/pandas/util/testing.py:22: DeprecationWarning: Importing from numpy.testing.decorators is deprecated since numpy 1.15.0, import from numpy.testing instead.
    from numpy.testing.decorators import slow     # noqa
/Library/Python/2.7/site-packages/pandas/util/nosetester.py:13
  /Library/Python/2.7/site-packages/pandas/util/nosetester.py:13: DeprecationWarning: Importing from numpy.testing.nosetester is deprecated since 1.15.0, import from numpy.testing instead.
    from numpy.testing import nosetester
-- Docs: https://docs.pytest.org/en/latest/warnings.html
!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!!
================ 3 skipped, 2 warnings, 1 error in 7.62 seconds ================
The command "sudo -H bash -x -e .travis/wheel.test.sh" exited with 2.
```

On Travis CI we don't really test Linux tests (they are covered by Google CI which likely will fail as well.

You can fix the above test failure to give it another try.

yongtang commented on tests/test_json.py:
```diff
@@ -0,0 +1,56 @@
+      label_filename,
+  )
+
+  d_train = data_ops.Dataset.zip((
```
I think you could use Tensorflow's public API `tf.data.Dataset.zip` instead.

jiachengxu replied:
Done

yongtang commented on tests/test_json.py:
```diff
@@ -0,0 +1,56 @@
+import os
+import tensorflow as tf
+import tensorflow_io.json as json_io
+from tensorflow_io.core.python.ops import data_ops as data_ops # pylint: disable=wrong-import-position
```
And remove this line (see using TensorFlow's public API `tf.data.Dataset.zip`).

jiachengxu replied:
Done

yongtang commented on tests/test_json_eager.py:
```diff
@@ -0,0 +1,84 @@
+  tf.compat.v1.enable_eager_execution()
+
+import tensorflow_io.json as json_io  # pylint: disable=wrong-import-position
+from tensorflow_io.core.python.ops import data_ops as data_ops # pylint: disable=wrong-import-position
```
Same here.

jiachengxu replied:
Done

yongtang commented on tests/test_json_eager.py:
```diff
@@ -0,0 +1,84 @@
+    i += 1
+  assert i == len(y_test)
+
+  dataset = data_ops.Dataset.zip((
```
same here.

jiachengxu replied:
Done

yongtang commented on tensorflow_io/json/python/ops/json_ops.py:
```diff
@@ -0,0 +1,63 @@
+  dataset = []
+  for sample in data:
+    sampledata = []
+    for value in sample.values():
```
@jiachengxu You also need to specify `columns` in order to correctly get the value in a defined order. Otherwise the `sample.values()` are non-deterministic. (Python 3 is failing because of that).

jiachengxu replied:
@yongtang , thanks for your review, I use the `sorted` to sort the key of the dict, I think this should solve the problem

yongtang commented:
@jiachengxu We are almost there for this step, but you need to specify the columns in order to have a deterministic order of the values. You can check `Ubuntu Python3 ` and see the failing test.

yongtang commented:
@jiachengxu we are almost there and all tests passes. However, one more thing is that, JSONDataset should pass a list `columns=[]` to specify the columns to be used, and the order of the list `columns` should be the rendered output order. for example, `columns=["floatfeature", "integerfeature"]` and `columns=["integerfeature", "floatfeature"]` should have different output. Can you update the PR to include `columns` as an argument?

jiachengxu commented:
Hi @yongtang, I added the `columns` as an argument of the JSONParser, if the `columns` is not specified, the Parser will just take all the cols in the dataset, and if the `columns` is specified, the output will respect the order and  the cols in the `columns`, I also added a test case to test that.

yongtang commented:
PR looks good to me 👍 . This PR could be merged first as the phase 1 of this project.

jiachengxu commented:
Hi @yongtang, I updated the WIP PR to Open, and also updated the description.

terrytangyuan commented on tensorflow_io/json/__init__.py:
```diff
@@ -0,0 +1,32 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""JSONInput
```
This should be JSONDataset?

jiachengxu replied:
Hi @terrytangyuan, yes, it should be JSONDataset, I have updated the PR
